src/Raven.Server/Documents/TransactionOperationsMerger.cs

```diff
@@ -848,5 +1092,173 @@ public void Dispose()
+                {
+                    using (item)
+                    {
+                        if (item.TryGet(TypeKey, out string strType) == false)
+                            continue;
+
+                        if (strType == StartRecordingType)
+                            continue;
+
+                        if (Enum.TryParse<TxInstruction>(strType, true, out var type))
+                        {
+                            switch (type)
+                            {
+                                case TxInstruction.BeginTx:
+                                    txDisposable = _parent.DocumentsStorage.ContextPool.AllocateOperationContext(out txCtx);
+                                    txCtx.OpenWriteTransaction();
+                                    break;
+                                case TxInstruction.Commit:
+                                    txCtx.Transaction.Commit();
+                                    break;
+                                case TxInstruction.Rollback:
+                                    break;
+                                case TxInstruction.DisposeTx:
+                                    txDisposable.Dispose();
+                                    break;
+                                case TxInstruction.FinishAsyncAndStartNew:
+                                    var previousTx = txCtx.Transaction;
+                                    txCtx.Transaction = txCtx.Transaction.BeginAsyncCommitAndStartNewTransaction();
+                                    txDisposable = txCtx.Transaction;
+
+                                    previousTx.EndAsyncCommit();
+                                    previousTx.Dispose();
+                                    break;
+                                case TxInstruction.EndAsyncCommit:
+                                    //Todo I think this is not relevant all the async instruction can run as one unit in  FinishAsyncAndStartNew
+                                    //previousTx.EndAsyncCommit();
+                                    break;
+                            }
+                            continue;
+                        }
+
+                        var cmd = ReadCommand(strType, item, context);
+
+                        cmd?.Execute(txCtx, null);
+                    }
+                }
+            }
+        }
+
+        private MergedTransactionCommand ReadCommand(string type, BlittableJsonReaderObject wrapCmdReader, JsonOperationContext context)
+        {
+            if (!wrapCmdReader.TryGet(CommandKey, out BlittableJsonReaderObject commandReader))
+            {
+                return null;
+                //Todo To remove after all MergedTransactionCommand driven class include in switch
+                throw new Exception("Can't read MergedTransactionCommand for replay");
+            }
+            MergedTransactionCommand cmd;
+            switch (type)
```

Comment:
Need to add a test to verify that all commands are known here

Reply:
In progress

Reply:
In progress